File: src/main/java/com/example/reactive/chapter1/completable/OrderService.java
package com.example.reactive.chapter1.completable;

import java.util.concurrent.CountDownLatch;

import com.example.reactive.chapter1.commons.Input;

import org.springframework.util.StopWatch;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RequiredArgsConstructor
public class OrderService {

    private final ShoppingCartService scService;
    private static CountDownLatch countDownLatch = new CountDownLatch(2);

    private void process() {

        scService.calculate(new Input())
                .thenAccept(output -> {
                    log.info("Complete!");
                    countDownLatch.countDown();
                });
    }

    public static void main(String[] args) throws InterruptedException {

        StopWatch stopWatch = new StopWatch();
        stopWatch.start();

        new OrderService(new CompletableFutureShoppingCartService()).process();
        new OrderService(new CompletableFutureShoppingCartService()).process();

//        countDownLatch.await();

        stopWatch.stop();
        log.info("took ... : {} ms", stopWatch.getTotalTimeMillis());
    }
}

File: src/main/kotlin/wmp/edu/reactive/store/callback/OrderService.kt
package wmp.edu.reactive.store.callback

import org.springframework.util.StopWatch
import wmp.edu.reactive.store.common.Input
import wmp.edu.reactive.store.common.Output
import wmp.edu.reactive.store.future.OrderService
import java.util.concurrent.CountDownLatch

val countDownLatch = CountDownLatch(2)

class OrderService {
    fun process(input: Input) {
        CallbackShoppingCart().calculate(input) {
            countDownLatch.countDown()
            Output("callback!")
        }
    }
}

fun main() {
    val stopWatch = StopWatch()
    stopWatch.start()

    OrderService().process(Input(1L))
    OrderService().process(Input(2L))

    wmp.edu.reactive.store.future.countDownLatch.await()
    stopWatch.stop()
    println(stopWatch.lastTaskTimeMillis)
}
